File: 05 - Smallest Multiple/python/main.py
import math

def smallestMultiple(upTo):
    # sqrt(10) = 3.16
    # up to 10: 2^3 * 3^2 * 5 * 7 = 2520
    #           p[i] ^ a[i] ...
    # p^a = k
    # a log(p) = log(k)
    # a = log(k) / log(p)
    result = 1
    continueChecking = True
    limit = math.sqrt(upTo) # no point checking higher
    primes = generatePrimes(upTo)
    exponents = [0]
    for i, val in enumerate(primes):
        exponents.append(1)
        if continueChecking:
            if primes[i] < limit:
                exponents[i] = math.floor(math.log(upTo) / math.log(primes[i]))
            else:
                continueChecking = False # dont check future exponents if p^2 > sqrt(upTo)
        result *= math.pow(primes[i], exponents[i])
        print("{} ^ {} = {} *".format(primes[i], exponents[i], result))
    return int(result)

def generatePrimes(upTo):
    composites = []
    primes = []
    for i in range (2, 7+1): # generate all composite numbers
        for j in range (2*i, upTo+1, i):
            composites.append(j)
    for i in range (2, upTo+1): # get all numbers not in composites
        if i not in composites:
            primes.append(i)
    return primes

result = smallestMultiple(5)
print("result: " + str(result))

File: 05 - Smallest Multiple/cpp/main.cpp
#include <vector>
#include <cstdint>
#include <algorithm>
#include <iostream>
#include <math.h>

std::vector<int64_t> generatePrimes(int64_t upTo) {
    std::vector<int64_t> composites = {};
    std::vector<int64_t> primes = {};
    for (int i = 2; i<= 7; i++) 
    {
        for (int j = 2 * i; j <= upTo; j += i)
        {
            primes.push_back(j);
        }
    }
    for (int i = 2; i <= upTo; i++)
    {
        if (std::find(primes.begin(), primes.end(), i) == primes.end())
        {
            composites.push_back(i);
        }
    }
    return composites;
}

int64_t smallestMultiple(int64_t upTo) {
    bool continueChecking = true;
    int64_t result = 1;
    int64_t limit = pow(upTo, 0.5) + 1;
    std::vector<int64_t> primes = generatePrimes(upTo);
    std::vector<int64_t> exponents = {};

    int i = 0;
    while (i < primes.size())
    {
        exponents.push_back(1);
        if (continueChecking) 
        {
            if (primes[i] < limit)
            {
                exponents[i] = std::floor(std::log(upTo) / std::log(primes[i]));
            }
            else
            {
                continueChecking = false;
            }
        }
        result *= pow(primes[i], exponents[i]);
        std::cout << primes[i] << " ^ " << exponents[i] << " = " << result << " *" << std::endl;
        i++;
    }
    return result;
}

int main() 
{
    int64_t result = smallestMultiple(20);
    std::cout << "result: " << result << std::endl;
}